source/lib/omnitrace/library/roctracer.cpp

```diff
@@ -913,17 +950,24 @@ hip_activity_callback(const char* begin, const char* end, void*)
             if(_kernel_names.find(_name) == _kernel_names.end())
                 _kernel_names.emplace(_name, tim::demangle(_name));
 
+            const auto _queue_track = get_perfetto_roctracer_per_stream()
+                                          ? perfetto::Track(queue_uuid(_queue))
+                                          : perfetto::ThreadTrack::Current();
+
             assert(_end_ns >= _beg_ns);
-            tracing::push_perfetto_ts(
-                category::device_hip{}, _kernel_names.at(_name).c_str(), _beg_ns,
-                perfetto::Flow::ProcessScoped(_cid), "begin_ns", _beg_ns, "corr_id",
-                record->correlation_id, "device", _devid, "queue", _queid, "pid",
-                record->process_id, "tid", _tid, "op", _op_id_names.at(record->op));
-            tracing::pop_perfetto_ts(category::device_hip{}, "", _end_ns, "end_ns",
-                                     _end_ns);
             // for some reason, this is necessary to make sure very last one ends
-            tracing::pop_perfetto_ts(category::device_hip{}, "", _end_ns, "end_ns",
-                                     _end_ns);
+            tracing::pop_perfetto_track(category::device_hip{}, "", _queue_track,
```

Comment:
This is very annoying, but somehow roctracer just reports incorrect start / end times, having this extra end **after** the trace means that sometimes kernels on the same stream look as if they overlap.
Ending before means that any unended event will be cut short by the next one.

Reply:
This is probably related to a roctracer bug I have seen elsewhere... You may have noticed, I have code that protects against roctracer returning end times that are before the start time. I recently had to patch a limiter into how many times that warning gets emitted bc it was showing up thousands of times in another app which was only running for a minute or two.

I am working on something else that is going to require control over the low-level HSA queues, at which point, I highly doubt there is going to be _any_ justification to continue to use either the roctracer or rocprofiler libraries for this information. Once I effectively have an internal implementation of roctracer/rocprofiler, I'll be able to get rid of this behavior.

Reply:
This was resolved by flushing perfetto before exiting the function